pkg/sdn/plugin/pod.go

```diff
@@ -0,0 +1,193 @@
+			},
+			Routes: []cnitypes.Route{
+				{
+					Dst: net.IPNet{
+						IP:   net.IPv4zero,
+						Mask: net.IPMask(net.IPv4zero),
+					},
+					GW: netutils.GenerateDefaultGateway(nodeNet),
+				},
+				{Dst: *clusterNetwork},
+			},
+		},
+	})
+}
+
+// Start the CNI server and start processing requests from it
+func (m *podManager) Start(socketPath string) error {
+	go m.processCNIRequests()
+
+	m.cniServer = cniserver.NewCNIServer(socketPath)
+	return m.cniServer.Start(m.handleCNIRequest)
+}
+
+// Returns a key for use with the runningPods map
+func getPodKey(request *cniserver.PodRequest) string {
+	return fmt.Sprintf("%s/%s", request.PodNamespace, request.PodName)
+}
+
+func (m *podManager) getPod(request *cniserver.PodRequest) *kubehostport.RunningPod {
+	return m.runningPods[getPodKey(request)]
+}
+
+// Return a list of Kubernetes RunningPod objects for hostport operations
+func (m *podManager) getRunningPods() []*kubehostport.RunningPod {
+	pods := make([]*kubehostport.RunningPod, 0)
+	for _, runningPod := range m.runningPods {
+		pods = append(pods, runningPod)
+	}
+	return pods
+}
+
+// Add a request to the podManager CNI request queue
+func (m *podManager) addRequest(request *cniserver.PodRequest) {
+	m.requests <- request
+}
+
+// Wait for and return the result of a pod request
+func (m *podManager) waitRequest(request *cniserver.PodRequest) *cniserver.PodResult {
+	return <-request.Result
+}
+
+// Enqueue incoming pod requests from the CNI server, wait on the result,
+// and return that result to the CNI client
+func (m *podManager) handleCNIRequest(request *cniserver.PodRequest) ([]byte, error) {
+	glog.V(5).Infof("Dispatching pod network request %v", request)
+	m.addRequest(request)
+	result := m.waitRequest(request)
+	glog.V(5).Infof("Returning pod network request %v, result %s err %v", request, string(result.Response), result.Err)
+	return result.Response, result.Err
+}
+
+// Process all CNI requests from the request queue serially.  Our OVS interaction
+// and scripts currently cannot run in parallel, and doing so greatly complicates
+// setup/teardown logic
+func (m *podManager) processCNIRequests() {
+	for request := range m.requests {
+		pk := getPodKey(request)
+
+		glog.V(5).Infof("Processing pod network request %v", request)
+		result := &cniserver.PodResult{}
+		switch request.Command {
+		case cniserver.CNI_ADD:
+			ipamResult, runningPod, err := m.podHandler.setup(request)
+			if ipamResult != nil {
+				result.Response, err = json.Marshal(ipamResult)
+				if result.Err == nil {
+					m.runningPods[pk] = runningPod
+				}
+			}
+			if err != nil {
+				result.Err = err
+			}
+		case cniserver.CNI_UPDATE:
+			result.Err = m.podHandler.update(request)
+		case cniserver.CNI_DEL:
+			delete(m.runningPods, pk)
+			result.Err = m.podHandler.teardown(request)
+		}
+		glog.V(5).Infof("Processed pod network request %v, result %s err %v", request, string(result.Response), result.Err)
+		request.Result <- result
+	}
```

Comment:
Unless you're adding a Close method, you should panic if you get here because that means this process is no longer managing pod updates (which is bad)